pvlib/spectrum.py

```diff
@@ -0,0 +1,366 @@
+    under cloudless skies. SPECTRL2 models radiative transmission, absorption,
+    and scattering due to atmospheric aerosol, water vapor, and ozone content.
+
+    Parameters
+    ----------
+    apparent_zenith : numeric
+        Solar zenith angle [degrees]
+    aoi : numeric
+        Angle of incidence of the solar vector on the panel [degrees]
+    surface_tilt : numeric
+        Panel tilt from horizontal [degrees]
+    ground_albedo : numeric
+        Albedo [0-1] of the ground surface. Can be provided as a scalar value
+        if albedo is not spectrally-dependent, or as a 122xN matrix where
+        the first dimension spans the wavelength range and the second spans
+        the number of simulations. [unitless]
+    surface_pressure : numeric
+        Surface pressure [Pa]
+    relative_airmass : numeric
+        Relative airmass. The airmass model used in [1]_ is the `'kasten1966'`
+        model, while a later implementation by NREL uses the
+        `'kastenyoung1989'` model. [unitless]
+    precipitable_water : numeric
+        Atmospheric water vapor content [cm]
+    ozone : numeric
+        Atmospheric ozone content [atm-cm]
+    aerosol_turbidity_500nm : numeric
+        Aerosol turbidity at 500 nm [unitless]
+    dayofyear : numeric, optional
+        The day of year [1-365].  Must be provided if ``apparent_zenith`` is
+        not a pandas Series.
+    scattering_albedo_400nm : numeric, default 0.945
+        Aerosol single scattering albedo at 400nm. The default value of 0.945
+        is suggested in [1]_ for a rural aerosol model. [unitless]
+    alpha : numeric, default 1.14
+        Angstrom turbidity exponent. The default value of 1.14 is suggested
+        in [1]_ for a rural aerosol model. [unitless]
+    wavelength_variation_factor : numeric, default 0.095
+        Wavelength variation factor [unitless]
+    aerosol_asymmetry_factor : numeric, default 0.65
+        Aeorosol asymmetry factor (mean cosine of scattering angle) [unitless]
+
+    Returns
+    -------
+    spectra : dict
+        A dict of arrays.  With the exception of `wavelength`, which has length
+        122, each array has shape (122, N) where N is the length of the
+        input ``apparent_zenith``.  All values are spectral irradiance
+        with units W/m^2/nm except for `wavelength`, which is in nanometers.
+
+            * wavelength
+            * dni_extra
+            * dhi
+            * dni
+            * poa_sky_diffuse
+            * poa_ground_diffuse
+            * poa_direct
+            * poa_global
+
+    Notes
+    -----
+    NREL's C implementation ``spectrl2_2.c`` [2]_ of the model differs in
+    several ways from the original report [1]_.  The report itself also has
+    a few differences between the in-text equations and the code appendix.
+    The list of known differences is shown below.  Note that this
+    implementation follows ``spectrl2_2.c``.
```

Comment:
I didn't compare with the Excel version in any meaningful way.  I agree it would be nice to be able to switch between the two (if there is indeed a difference), but I hesitate to add additional complexity to this PR.  Future work?

Reply:
maybe someone at NREL would know if the C code or the Excel version is in some sense authoritative. I agree with deferring any kind of switch.